src/uct/tcp/tcp_ep.c

```diff
@@ -730,6 +748,58 @@ static inline unsigned uct_tcp_ep_recv(uct_tcp_ep_t *ep, size_t recv_length)
     return 1;
 }
 
+void uct_tcp_ep_magic_number_progress_rx(uct_tcp_ep_t *ep)
+{
+    uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
+                                            uct_tcp_iface_t);
+    char str_local_addr[UCS_SOCKADDR_STRING_LEN];
+    char str_remote_addr[UCS_SOCKADDR_STRING_LEN];
+    ucs_status_t status;
+    size_t recv_length;
+
+    if (ep->rx.buf == NULL) {
+        /* Don't allocate from iface's RX memory pool to avoid data races */
+        ep->rx.buf = &ep->magic_number;
```

Comment:
maybe would be simpler to read the magic number into rx buf and later compare it to the the known value ?